File: Medium/22. Generate Parentheses/solution.cpp
class Solution {
public:
    vector<string> result;
    string s = "";
    void generate(int cntOpen, int cntClose, int n, string s){
        if (cntOpen == cntClose && cntOpen == n){
            result.push_back(s);
            return;
        }
        if (cntOpen < n){
            s.push_back('(');
            generate(cntOpen + 1, cntClose, n, s);
            s.pop_back();
        }
        if (cntClose < cntOpen){
            s.push_back(')');
            generate(cntOpen, cntClose + 1, n, s);
            s.pop_back();
        }
    }
    vector<string> generateParenthesis(int n) {
        generate(0, 0, n, s);
        return result;
    }
};

File: Medium/22. Generate Parentheses/solution.py
class Solution:
    def helper(self, o, c, n, p, ret):
        if o == n and c == n:
            ret.append(p)

        if o < n:
            self.helper(o + 1, c, n, p + "(", ret)
        
        if c < o:
            self.helper(o, c + 1, n, p + ")", ret)

    
    def generateParenthesis(self, n: int) -> List[str]:
        ret = []
        self.helper(0, 0, n, "", ret)

        return ret

        